update servo [2015-12-02]

vyp opened the pull request:
1. Obviously don't merge, this does not build.
2. Don't know any rust -> **don't know what I'm doing**.
3. Will squash commits if you want. Put multiple for now because I think it's easier to review?
4. The error I told you about yesterday (http://pastebin.com/h1DSY85L) seems to have been fixed before I could track it down, so now cargo actually gets to the compilation stage.
5. I changed the source for `glutin` in Cargo.toml to just `"*"` because I guess the `{ git = ... }` was there because glutin wasn't a crate at that time?
6. Of course, feel free to disregard this pr entirely and/or otherwise fix the remaining errors yourself if you already know how!

After getting the latest Cargo.lock from servo/servo/components/servo/Cargo.lock, the `cargo build` error is:

```
src/util.rs:21:2: 21:21 error: attempted access of field `resources_path` on type `util::opts::Opts`, but no field with that name was found
src/util.rs:21  opts.resources_path = None;
                ^~~~~~~~~~~~~~~~~~~
src/util.rs:21:7: 21:21 help: did you mean `userscripts`?
src/util.rs:21  opts.resources_path = None;
                     ^~~~~~~~~~~~~~
src/browser/browser.rs:78:21: 78:47 error: mismatched types:
 expected `*const libc::c_void`,
    found `*const ()`
(expected enum `libc::c_void`,
    found ()) [E0308]
src/browser/browser.rs:78       gl::load_with(|s| window.get_proc_address(s));
                                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~
src/browser/browser.rs:78:21: 78:47 help: run `rustc --explain E0308` to see a detailed explanation
src/browser/buffer.rs:47:23: 52:28 error: this function takes 1 parameter but 9 parameters were supplied [E0061]
src/browser/buffer.rs:47        let constellation = Constellation::<LayoutTask, ScriptTask>::start(
src/browser/buffer.rs:48            proxy.clone_compositor_proxy(),
src/browser/buffer.rs:49            browser.resource(),
src/browser/buffer.rs:50            browser.cache().image(), browser.cache().font(),
src/browser/buffer.rs:51            time_profiler.clone(), mem_profiler.clone(), None,
src/browser/buffer.rs:52            browser.storage(), true);
src/browser/buffer.rs:47:23: 52:28 help: run `rustc --explain E0061` to see a detailed explanation
src/browser/buffer.rs:54:24: 55:47 error: this function takes 2 parameters but 6 parameters were supplied [E0061]
src/browser/buffer.rs:54        let mut compositor = CompositorTask::create(Some(window.clone()), proxy, receiver,
src/browser/buffer.rs:55            constellation, time_profiler, mem_profiler);
src/browser/buffer.rs:54:24: 55:47 help: run `rustc --explain E0061` to see a detailed explanation
src/browser/buffer.rs:115:19: 115:29 error: no method named `shutdown` found for type `Box<compositing::compositor_task::CompositorEventListener + 'static>` in the current scope
src/browser/buffer.rs:115       self.compositor.shutdown();
                                                ^~~~~~~~~~
src/main.rs:48:21: 48:47 error: mismatched types:
 expected `*const libc::c_void`,
    found `*const ()`
(expected enum `libc::c_void`,
    found ()) [E0308]
src/main.rs:48      gl::load_with(|s| window.get_proc_address(s));
                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~
src/main.rs:48:21: 48:47 help: run `rustc --explain E0308` to see a detailed explanation
error: aborting due to 6 previous errors
Could not compile `miserve`.
```

(All the dependencies build fine, with rustc `1805bba399805d8fd8e85e23c31a0580f21533cb`.)

I've only fixed the first one (I think!), and now I'm already stuck on the 5th error. -_- The `shutdown` method was removed in https://github.com/servo/servo/pull/7913. As you can see in that diff, it seems to have been replaced with the `start/finish_shutting_down` functions, but I'm not sure how to use/access them. Because even though they have `pub`, it doesn't look like the `compositor` module has been made publicly accessible: https://github.com/servo/servo/blob/2be60be062e14c937af601faed78a6aceccdb062/components/compositing/lib.rs#L58 ?

I'm guessing this is probably just down to my lack of knowledge about rust right now..., hence why I've decided to open this prematurely. I would have preferred to have fixed everything and have it all working before opening, but I realised that's harder than I expected... so opening a wip pr might make it a little easier, if you or others can find time. Speaking of which, no rush, it's (close to?) holiday season.

vyp committed:
Cargo.toml: use simple semver expression for glutin's version number

Because it is a crate now.

vyp committed:
Cargo.lock: update to latest components/servo/Cargo.lock from servo

This compiles with rustc 1805bba399805d8fd8e85e23c31a0580f21533cb
(2015-11-26).

vyp committed:
util: no field with name `resources_path` exists under servo::util::opts

The compile time error:

    src/util.rs:21:2: 21:21 error: attempted access of field `resources_path` on type `util::opts::Opts`, but no field with that name was found
    src/util.rs:21 	opts.resources_path = None;
                    ^~~~~~~~~~~~~~~~~~~
    src/util.rs:21:7: 21:21 help: did you mean `userscripts`?
    src/util.rs:21 	opts.resources_path = None;
                         ^~~~~~~~~~~~~~

vyp committed:
browser: fix comment typo

meh commented:
You seem to have fixed the resources one yourself, am I right?

The `gl::load_with` call should be fixable with adding a `as *const _` at the end of the call inside the closure.

About the shutdown you should ask in #servo, they'll know better, maybe it shut downs automagically now.

Thanks for working on it :panda_face:

vyp committed:
fix mismatched types of `*const libc::c_void` vs `*const ()`

Where `*const libc::c_void` is expected but found type is `*const ()`.
Compile time errors:

    src/browser/browser.rs:78:21: 78:47 error: mismatched types:
     expected `*const libc::c_void`,
        found `*const ()`
    (expected enum `libc::c_void`,
        found ()) [E0308]
    src/browser/browser.rs:78 		gl::load_with(|s| window.get_proc_address(s));
                                                              ^~~~~~~~~~~~~~~~~~~~~~~~~~
    src/browser/browser.rs:78:21: 78:47 help: run `rustc --explain E0308` to see a detailed explanation

...

    src/main.rs:48:21: 48:47 error: mismatched types:
     expected `*const libc::c_void`,
        found `*const ()`
    (expected enum `libc::c_void`,
        found ()) [E0308]
    src/main.rs:48 		gl::load_with(|s| window.get_proc_address(s));
                                              ^~~~~~~~~~~~~~~~~~~~~~~~~~
    src/main.rs:48:21: 48:47 help: run `rustc --explain E0308` to see a detailed explanation

vyp commented:
> You seem to have fixed the resources one yourself, am I right?

As in, the compile error goes away. Problem is I couldn't find a `.resources_path` field for `servo::util::opts::Opts`, even in older versions of the `components/util/opts.rs` file. So I've simply replaced it with a call to `servo::util::resource_files::set_resources_path`, which by reading the function name (and type annotation), seems to do what you want. Another problem here being that I don't know if I'm breaking things here because I haven't run it yet (because it obviously doesn't compile). And also, because I couldn't compile the current version of miserve (on master), I don't know how it's supposed to work either.

Anyway, you're right about `as *const _`, which means just three errors are left.

vyp committed:
buffer: use InitialConstellationState struct for the constellation

Fixes the following compile time error:

    src/browser/buffer.rs:47:23: 52:28 error: this function takes 1 parameter but 9 parameters were supplied [E0061]
    src/browser/buffer.rs:47 		let constellation = Constellation::<LayoutTask, ScriptTask>::start(
    src/browser/buffer.rs:48 			proxy.clone_compositor_proxy(),
    src/browser/buffer.rs:49 			browser.resource(),
    src/browser/buffer.rs:50 			browser.cache().image(), browser.cache().font(),
    src/browser/buffer.rs:51 			time_profiler.clone(), mem_profiler.clone(), None,
    src/browser/buffer.rs:52 			browser.storage(), true);
    src/browser/buffer.rs:47:23: 52:28 help: run `rustc --explain E0061` to see a detailed explanation

All these separate arguments were combined into a
'InitialConstellationState' struct at [1].

[1]: https://github.com/servo/servo/pull/7645

vyp committed:
buffer: use InitialCompositorState struct for the compositor

Fixes the following compile time error:

    src/browser/buffer.rs:54:24: 55:47 error: this function takes 2 parameters but 6 parameters were supplied [E0061]
    src/browser/buffer.rs:54 		let mut compositor = CompositorTask::create(Some(window.clone()), proxy, receiver,
    src/browser/buffer.rs:55 			constellation, time_profiler, mem_profiler);
    src/browser/buffer.rs:54:24: 55:47 help: run `rustc --explain E0061` to see a detailed explanation

Arguments 2-6 were combined into a 'InitialCompositorState' struct at
[1].

[1]: https://github.com/servo/servo/pull/7645

vyp committed:
buffer: Drop trait for Buffer is no longer needed

Because compositor properly handles its own shutdown now (assuming
handle_events is called).

Fixes the following compile time error:

    src/browser/buffer.rs:115:19: 115:29 error: no method named `shutdown` found for type `Box<compositing::compositor_task::CompositorEventListener + 'static>` in the current scope
    src/browser/buffer.rs:115 		self.compositor.shutdown();
                                                            ^~~~~~~~~~

vyp committed:
main: result_expect is stable and std::thread::sleep_ms is deprecated

Addresses the following compiler warnings:

    src/main.rs:1:24: 1:37 warning: this feature is stable. attribute no longer needed, #[warn(stable_features)] on by default
    src/main.rs:1 #![feature(box_syntax, result_expect)]
                                         ^~~~~~~~~~~~~
    src/main.rs:75:4: 75:20 warning: use of deprecated item: replaced by `std::thread::sleep`, #[warn(deprecated)] on by default
    src/main.rs:75 			thread::sleep_ms(10);
                                    ^~~~~~~~~~~~~~~~

vyp commented:
1. Sorry for the delay here, I decided I needed to learn rust just to more easily make sense of it all. So the past week or so I've been (slowly) going though the rust book.
2. Would you like me to squash?
3. You were right about the shutdown. :)
4. I do notice that Servo now has a new Cargo.lock. Not sure how much you care about keeping up to date with the latest Servo all the time. But if you do want to update to it, I suggest doing so from this PR because I expect the changes will be smaller/easier due to the smaller time difference.
   - My reason for this PR is because I couldn't build (current) master, but now that I can (if this is merged), I don't really mind about _always_ keeping up to date with servo.
   - If you do want to, but don't have the time, I can open another PR to start work on it if you want?

---

Now building is one thing, but maybe I have broken how it actually runs. When I run `./target/debug/miserve`, I get a panic:

```
[glutin] x error code=8 major=42 minor=0!  [20:53]
thread '<main>' panicked at 'no url given', src/libcore/option.rs:335
```

So maybe this PR is not actually ready yet.. did you get this before (i.e. on current master)? (And if so, what is the correct way to run miserve then?)

I think this might be something to do with setting the initial url field on the `servo::util::opts` module, so I tried doing:

``` rust
opts.url = Url::parse("https://duckduckgo.com").ok();
```

in `util.rs` (instead of `opts.url = None;`), but get the same panic. I will try to dig deeper when I get the time, but any ideas?

vyp commented:
Btw if it helps anyone, I believe now that the first line of the above error is actually a separate error or warning than the second panic line. (I still haven't been able to track down the source of the panic, mainly due to lack of time.)
